tests/test_edgeql_syntax.py

```diff
@@ -3356,49 +3330,49 @@ def test_edgeql_syntax_selectfor_05(self):
         """
 
     @tb.must_fail(errors.EdgeQLSyntaxError,
-                  'Missing parentheses around complex expression in a '
-                  'FOR iterator clause',
+                  'Unexpected keyword \'.*\'',
                   hint=None, line=2, col=18)
     def test_edgeql_syntax_selectfor_06(self):
         """
         FOR x in DETACHED foo UNION x;
         """
+        # XXX: error recovery quality regression
 
     @tb.must_fail(errors.EdgeQLSyntaxError,
-                  'Missing parentheses around complex expression in a '
-                  'FOR iterator clause',
-                  hint=None, line=2, col=18)
+                  'Missing keyword \'.*\'',
+                  hint=None, line=2, col=21)
     def test_edgeql_syntax_selectfor_08(self):
         """
         FOR x in foo + bar UNION x;
         """
+        # XXX: error recovery quality regression
 
     @tb.must_fail(errors.EdgeQLSyntaxError,
-                  'Missing parentheses around complex expression in a '
-                  'FOR iterator clause',
-                  hint=None, line=2, col=18)
+                  'Missing keyword \'.*\'',
+                  hint=None, line=2, col=25)
     def test_edgeql_syntax_selectfor_09(self):
         """
         FOR x in foo.bar + bar UNION x;
         """
+        # XXX: error recovery quality regression
 
     @tb.must_fail(errors.EdgeQLSyntaxError,
-                  'Missing parentheses around complex expression in a '
-                  'FOR iterator clause',
-                  hint=None, line=2, col=18)
+                  'Missing keyword \'.*\'',
+                  hint=None, line=2, col=21)
     def test_edgeql_syntax_selectfor_10(self):
         """
         FOR x in foo { x } UNION x;
         """
+        # XXX: error recovery quality regression
 
     @tb.must_fail(errors.EdgeQLSyntaxError,
-                  'Missing parentheses around complex expression in a '
-                  'FOR iterator clause',
+                  'Unexpected keyword \'.*\'',
                   hint=None, line=2, col=18)
     def test_edgeql_syntax_selectfor_11(self):
         """
         FOR x in SELECT 1 UNION x;
         """
+        # XXX: error recovery quality regression
```

Comment:
It is nice to have an error thrown you change behavior somehow. If I change this to xfail and then change the parser to produce some third error here, it will not catch it.

Reply:
> How hard do you think it will be to repair this regression?

Not that hard, actually. But it would require special handling in the parser.